src/librustc/ty/sty.rs

```diff
@@ -2154,30 +2154,26 @@ impl<'tcx> Const<'tcx> {
     pub fn try_eval_bits(
         &self,
         tcx: TyCtxt<'_, '_, '_>,
-        ty: ParamEnvAnd<'_, Ty<'_>>,
+        param_env: ParamEnv<'_>,
+        ty: Ty<'_>,
     ) -> Option<u128> {
-        assert_eq!(self.ty, ty.value);
+        assert_eq!(self.ty, ty);
         let tcx = tcx.global_tcx();
-        let ty = if ty.value.needs_infer() || ty.value.has_placeholders() {
-            tcx
-                .lift_to_global(&ty)
-                .unwrap_or_else(|| bug!("cannot lift {:#?} for {:#?}", ty, self))
-        } else {
-            let ty = tcx
-                .lift_to_global(&ty.value)
-                .unwrap_or_else(|| bug!("cannot lift {:#?} for {:#?}", ty, self));
-            ParamEnv::empty().and(ty)
-        };
+        // if `ty` has no unresolved components, use an empty param_env
+        let ty = tcx.lift_to_global(&param_env.with_reveal_all().and(ty)).unwrap();
         let size = tcx.layout_of(ty).ok()?.size;
         match self.val {
             ConstValue::Unevaluated(did, substs) => {
-                let substs = tcx.lift_to_global(&substs).unwrap();
-                let instance = ty::Instance::resolve(tcx, ty.param_env, did, substs)?;
+                // if `substs` has no unresolved components, use and empty param_env
+                let pem_and_substs = param_env.with_reveal_all().and(substs);
+                let (param_env, substs) = tcx.lift_to_global(&pem_and_substs).unwrap().into_parts();
```

Comment:
IMO you should still fail gracefully instead of unwrapping.

Reply:
Won't that just shift the problem to the user of `try_eval_bits` and make it much harder to debug what's wrong?

Reply:
Resolved by getting rid of the lift entirely